pkg/controller/certificates/signer/signer_test.go

```diff
@@ -99,3 +103,156 @@ func TestSigner(t *testing.T) {
+
+func TestHandle(t *testing.T) {
+	cases := []struct {
+		name       string
+		approved   bool
+		signerName string
+		err        bool
+		verify     func(*testing.T, []testclient.Action)
+	}{
+		{
+			name:       "should sign if signerName is kubernetes.io/kube-apiserver-client",
+			signerName: "kubernetes.io/kube-apiserver-client",
+			approved:   true,
+			verify: func(t *testing.T, as []testclient.Action) {
+				if len(as) != 1 {
+					t.Errorf("expected one Update action but got %d", len(as))
+					return
+				}
+				csr := as[0].(testclient.UpdateAction).GetObject().(*capi.CertificateSigningRequest)
+				if len(csr.Status.Certificate) == 0 {
+					t.Errorf("expected certificate to be issued but it was not")
+				}
+			},
+		},
+		{
+			name:       "should sign if signerName is kubernetes.io/kube-apiserver-client-kubelet",
+			signerName: "kubernetes.io/kube-apiserver-client-kubelet",
+			approved:   true,
+			verify: func(t *testing.T, as []testclient.Action) {
+				if len(as) != 1 {
+					t.Errorf("expected one Update action but got %d", len(as))
+					return
+				}
+				csr := as[0].(testclient.UpdateAction).GetObject().(*capi.CertificateSigningRequest)
+				if len(csr.Status.Certificate) == 0 {
+					t.Errorf("expected certificate to be issued but it was not")
+				}
+			},
+		},
+		{
+			name:       "should sign if signerName is kubernetes.io/legacy-unknown",
+			signerName: "kubernetes.io/legacy-unknown",
+			approved:   true,
+			verify: func(t *testing.T, as []testclient.Action) {
+				if len(as) != 1 {
+					t.Errorf("expected one Update action but got %d", len(as))
+					return
+				}
+				csr := as[0].(testclient.UpdateAction).GetObject().(*capi.CertificateSigningRequest)
+				if len(csr.Status.Certificate) == 0 {
+					t.Errorf("expected certificate to be issued but it was not")
+				}
+			},
+		},
+		{
+			name:       "should sign if signerName is kubernetes.io/kubelet-serving",
+			signerName: "kubernetes.io/kubelet-serving",
+			approved:   true,
+			verify: func(t *testing.T, as []testclient.Action) {
+				if len(as) != 1 {
+					t.Errorf("expected one Update action but got %d", len(as))
+					return
+				}
+				csr := as[0].(testclient.UpdateAction).GetObject().(*capi.CertificateSigningRequest)
+				if len(csr.Status.Certificate) == 0 {
+					t.Errorf("expected certificate to be issued but it was not")
+				}
+			},
+		},
+		{
+			name:       "should do nothing if an unrecognised signerName is used",
+			signerName: "kubernetes.io/not-recognised",
+			approved:   true,
+			verify: func(t *testing.T, as []testclient.Action) {
+				if len(as) != 0 {
+					t.Errorf("expected no action to be taken")
+				}
+			},
+		},
+		{
+			name:       "should do nothing if not approved",
+			signerName: "kubernetes.io/kubelet-serving",
+			verify: func(t *testing.T, as []testclient.Action) {
+				if len(as) != 0 {
+					t.Errorf("expected no action to be taken")
+				}
+			},
+		},
+		{
+			name: "should do nothing if signerName is nil",
```

Comment:
This should panic right?

Reply:
(if the CSR was approved it should deference the `nil` and `panic`)

Reply:
Correct - but it's now no longer possible/valid to pass a CSR to `handle` that sets a `nil` signerName (or at least, it is unsupported/unhandled). So I'm not sure if this is really a valid test case to include anymore, unless we add a check into `handle` for a nil signerName and return early explicitly?

With the CSR being *not* approved, it is also expected that this test case does nothing, although it is then effectively a dupe of the case above it (`should do nothing if not approved`).

Reply:
Updated these tests. We no longer have a test with an explicitly `nil` signerName (and do not expect/support a CSR with a nil signerName to be passed to handle). I've also expanded out the test coverage here too with a few additional cases (as well as fixing the failures)

Reply:
Not testing for `nil` is fine.  Since the code would `panic`, the worse case is a DOS against the signer.